foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/PoetryMigration.java

```diff
@@ -0,0 +1,118 @@
+ * #L%
+ */
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Optional;
+import java.util.Set;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.boozallen.aissemble.upgrade.util.FileUtils;
+import com.electronwill.nightconfig.core.Config;
+import com.electronwill.nightconfig.core.file.FileConfig;
+
+/**
+ * 
+ * Baton migration used to migrate aiSSEMBLE python packages to the new naming convention
+ */
+public class PoetryMigration extends AbstractAissembleMigration {
+
+    private static final Logger logger = LoggerFactory.getLogger(PoetryMigration.class);
+    protected static final String DEPENDENCIES_KEY = "tool.poetry.dependencies";
+    protected static final String PYTHON_PACKAGE_PREFIX = "aissemble-";
+    protected static final String UPGRADE_VERSION = "1.7.0.*";
+
+    protected static final Set<String> OLD_PYTHON_PACKAGES = Set.of(
+        "foundation-core-python", 
+        "foundation-pdp-client-python", 
+        "foundation-model-lineage", 
+        "foundation-encryption-policy-python", 
+        "extensions-encryption-vault-python", 
+        "extensions-data-delivery-spark-py", 
+        "foundation-data-lineage-python"
+    );
+
+    @Override
+    protected boolean shouldExecuteOnFile(File file) {
+        if(file == null || !file.exists()) {
+            logger.error("Unable to read file {} to check if migration should be executed", file.getAbsolutePath());
+        }
+        boolean shouldExecute = false;
+        FileConfig poetryConfig = FileConfig.of(file);
+        poetryConfig.load();
+        Optional<Config> dependenciesOpt = poetryConfig.getOptional(DEPENDENCIES_KEY);
+        if(dependenciesOpt.isPresent()) {
+            Set<String> dependencies = dependenciesOpt.get().valueMap().keySet();
+            shouldExecute = hasOldPackages(dependencies);
+        }
+        else {
+            logger.warn("Could not get dependencies for file {}", file.getAbsolutePath());
+        }
+        return shouldExecute;
+    }
+
+    private boolean hasOldPackages(Set<String> packages) {
+        boolean hasOldPackages = false;
+        for(String oldPackage : OLD_PYTHON_PACKAGES) {
+            if(packages.contains(oldPackage)) {
+                hasOldPackages = true;
+                break;
+            }
+        }
+        return hasOldPackages;
+    }
+
+    @Override
+    protected boolean performMigration(File file) {
+        if(file == null || !file.exists()) {
+            logger.error("Unable to read file {} for migration", file.getAbsolutePath());
+        }
+        boolean performedSuccessfully = false;
+        FileConfig pyproject = FileConfig.of(file);
+        pyproject.load();
+        Optional<Config> dependenciesOpt = pyproject.getOptional(DEPENDENCIES_KEY);
+        if(dependenciesOpt.isPresent()) {
+            try {
+                performedSuccessfully = migrateOldPythonPackages(dependenciesOpt.get(), file);
```

Comment:
S: Instead of reading in the file and getting the dependencies again, you get add an attribute to the class `private Config config;` that you set in your `shouldExecuteOnFile` and can then access in this functions. Would remove a lot of the duplicate code.

Reply:
I removed the Optional logic.